File: packages/ai/src/providers/transform-messages.ts
import type { Api, AssistantMessage, Message, Model, ToolCall, ToolResultMessage } from "../types.js";

/**
 * 标准化工具调用ID以实现跨提供者兼容性。
 * OpenAI Responses API生成的ID长度超过450个字符，包含特殊字符如`|`。
 * Anthropic API要求ID匹配^[a-zA-Z0-9_-]+$（最大64个字符）。
 */
export function transformMessages<TApi extends Api>(
	messages: Message[],
	model: Model<TApi>,
	normalizeToolCallId?: (id: string, model: Model<TApi>, source: AssistantMessage) => string,
): Message[] {
	// 构建原始工具调用ID到标准化ID的映射
	const toolCallIdMap = new Map<string, string>();

	// 第一遍：转换消息（思考块、工具调用ID标准化）
	const transformed = messages.map((msg) => {
		// 用户消息保持不变
		if (msg.role === "user") {
			return msg;
		}

		// 处理toolResult消息 - 如果有映射则标准化toolCallId
		if (msg.role === "toolResult") {
			const normalizedId = toolCallIdMap.get(msg.toolCallId);
			if (normalizedId && normalizedId !== msg.toolCallId) {
				return { ...msg, toolCallId: normalizedId };
			}
			return msg;
		}

		// 助手消息需要转换检查
		if (msg.role === "assistant") {
			const assistantMsg = msg as AssistantMessage;
			const isSameModel =
				assistantMsg.provider === model.provider &&
				assistantMsg.api === model.api &&
				assistantMsg.model === model.id;

			const transformedContent = assistantMsg.content.flatMap((block) => {
				if (block.type === "thinking") {
					// 对于相同模型：保留带签名的思考块（重放所需）
					// 即使思考文本为空（OpenAI加密推理）
					if (isSameModel && block.thinkingSignature) return block;
					// 跳过空的思考块，将其他转换为纯文本
					if (!block.thinking || block.thinking.trim() === "") return [];
					if (isSameModel) return block;
					return {
						type: "text" as const,
						text: block.thinking,
					};
				}

				if (block.type === "text") {
					if (isSameModel) return block;
					return {
						type: "text" as const,
						text: block.text,
					};
				}

				if (block.type === "toolCall") {
					const toolCall = block as ToolCall;
					let normalizedToolCall: ToolCall = toolCall;

					if (!isSameModel && toolCall.thoughtSignature) {
						normalizedToolCall = { ...toolCall };
						delete (normalizedToolCall as { thoughtSignature?: string }).thoughtSignature;
					}

					if (!isSameModel && normalizeToolCallId) {
						const normalizedId = normalizeToolCallId(toolCall.id, model, assistantMsg);
						if (normalizedId !== toolCall.id) {
							toolCallIdMap.set(toolCall.id, normalizedId);
							normalizedToolCall = { ...normalizedToolCall, id: normalizedId };
						}
					}

					return normalizedToolCall;
				}

				return block;
			});

			return {
				...assistantMsg,
				content: transformedContent,
			};
		}
		return msg;
	});

	// 第二遍：为孤立的工具调用插入合成的空工具结果
	// 这样可以保留思考签名并满足API要求
	const result: Message[] = [];
	let pendingToolCalls: ToolCall[] = [];
	let existingToolResultIds = new Set<string>();

	for (let i = 0; i < transformed.length; i++) {
		const msg = transformed[i];

		if (msg.role === "assistant") {
			// 如果我们有待处理的来自前一个助手的孤立工具调用，现在插入合成结果
			if (pendingToolCalls.length > 0) {
				for (const tc of pendingToolCalls) {
					if (!existingToolResultIds.has(tc.id)) {
						result.push({
							role: "toolResult",
							toolCallId: tc.id,
							toolName: tc.name,
							content: [{ type: "text", text: "未提供结果" }],
							isError: true,
							timestamp: Date.now(),
						} as ToolResultMessage);
					}
				}
				pendingToolCalls = [];
				existingToolResultIds = new Set();
			}

			// 完全跳过错误/中止的助手消息。
			// 这些是不完整的回合，不应该被重放：
			// - 可能有部分内容（只有推理没有消息、不完整的工具调用）
			// - 重放它们可能导致API错误（例如，OpenAI "reasoning without following item"）
			// - 模型应该从最后一个有效状态重新尝试
			const assistantMsg = msg as AssistantMessage;
			if (assistantMsg.stopReason === "error" || assistantMsg.stopReason === "aborted") {
				continue;
			}

			// 跟踪此助手消息中的工具调用
			const toolCalls = assistantMsg.content.filter((b) => b.type === "toolCall") as ToolCall[];
			if (toolCalls.length > 0) {
				pendingToolCalls = toolCalls;
				existingToolResultIds = new Set();
			}

			result.push(msg);
		} else if (msg.role === "toolResult") {
			existingToolResultIds.add(msg.toolCallId);
			result.push(msg);
		} else if (msg.role === "user") {
			// 用户消息中断工具流程 - 为孤立调用插入合成结果
			if (pendingToolCalls.length > 0) {
				for (const tc of pendingToolCalls) {
					if (!existingToolResultIds.has(tc.id)) {
						result.push({
							role: "toolResult",
							toolCallId: tc.id,
							toolName: tc.name,
							content: [{ type: "text", text: "未提供结果" }],
							isError: true,
							timestamp: Date.now(),
						} as ToolResultMessage);
					}
				}
				pendingToolCalls = [];
				existingToolResultIds = new Set();
			}
			result.push(msg);
		} else {
			result.push(msg);
		}
	}

	return result;
}

File: py-packages/ai/pi_ai/providers/transform_messages.py
"""
Message transformation utilities for cross-provider compatibility.
"""

from __future__ import annotations

import time
from typing import Callable

from pi_ai.types import (
    Api,
    AssistantMessage,
    Message,
    Model,
    TextContent,
    ToolCall,
    ToolResultMessage,
)


def transform_messages(
    messages: list[Message],
    model: Model,
    normalize_tool_call_id: Callable[[str, Model, AssistantMessage], str] | None = None,
) -> list[Message]:
    """
    Transform messages for cross-provider compatibility.
    
    Handles:
    - Thinking block conversion for different models
    - Tool call ID normalization
    - Synthetic tool results for orphaned tool calls
    - Filtering errored/aborted assistant messages
    
    Args:
        messages: List of messages to transform
        model: Target model
        normalize_tool_call_id: Optional function to normalize tool call IDs
    
    Returns:
        Transformed list of messages
    """
    # Build a map of original tool call IDs to normalized IDs
    tool_call_id_map: dict[str, str] = {}

    # First pass: transform messages
    transformed: list[Message] = []
    for msg in messages:
        # User messages pass through unchanged
        if msg.role == "user":
            transformed.append(msg)
            continue

        # Handle toolResult messages - normalize toolCallId if we have a mapping
        if msg.role == "toolResult":
            normalized_id = tool_call_id_map.get(msg.tool_call_id)
            if normalized_id and normalized_id != msg.tool_call_id:
                new_msg = msg.model_copy(update={"tool_call_id": normalized_id})
                transformed.append(new_msg)
            else:
                transformed.append(msg)
            continue

        # Assistant messages need transformation
        if msg.role == "assistant":
            assistant_msg = msg
            is_same_model = (
                assistant_msg.provider == model.provider
                and assistant_msg.api == model.api
                and assistant_msg.model == model.id
            )

            transformed_content = []
            for block in assistant_msg.content:
                if block.type == "thinking":
                    # For same model: keep thinking blocks with signatures
                    if is_same_model and block.thinking_signature:
                        transformed_content.append(block)
                    # Skip empty thinking blocks
                    elif not block.thinking or block.thinking.strip() == "":
                        continue
                    elif is_same_model:
                        transformed_content.append(block)
                    else:
                        # Convert to plain text for other models
                        transformed_content.append(
                            TextContent(type="text", text=block.thinking)
                        )

                elif block.type == "text":
                    if is_same_model:
                        transformed_content.append(block)
                    else:
                        transformed_content.append(
                            TextContent(type="text", text=block.text)
                        )

                elif block.type == "toolCall":
                    tool_call = block
                    normalized_tool_call = tool_call

                    # Remove thought signature for different models
                    if not is_same_model and tool_call.thought_signature:
                        normalized_tool_call = tool_call.model_copy(
                            update={"thought_signature": None}
                        )

                    # Normalize tool call ID
                    if not is_same_model and normalize_tool_call_id:
                        normalized_id = normalize_tool_call_id(
                            tool_call.id, model, assistant_msg
                        )
                        if normalized_id != tool_call.id:
                            tool_call_id_map[tool_call.id] = normalized_id
                            normalized_tool_call = normalized_tool_call.model_copy(
                                update={"id": normalized_id}
                            )

                    transformed_content.append(normalized_tool_call)

                else:
                    transformed_content.append(block)

            new_assistant_msg = assistant_msg.model_copy(
                update={"content": transformed_content}
            )
            transformed.append(new_assistant_msg)
            continue

        # Pass through unknown message types
        transformed.append(msg)

    # Second pass: insert synthetic empty tool results for orphaned tool calls
    result: list[Message] = []
    pending_tool_calls: list[ToolCall] = []
    existing_tool_result_ids: set[str] = set()

    for msg in transformed:
        if msg.role == "assistant":
            # If we have pending orphaned tool calls, insert synthetic results
            if pending_tool_calls:
                for tc in pending_tool_calls:
                    if tc.id not in existing_tool_result_ids:
                        result.append(
                            ToolResultMessage(
                                role="toolResult",
                                tool_call_id=tc.id,
                                tool_name=tc.name,
                                content=[TextContent(type="text", text="No result provided")],
                                is_error=True,
                                timestamp=int(time.time() * 1000),
                            )
                        )
                pending_tool_calls = []
                existing_tool_result_ids = set()

            # Skip errored/aborted assistant messages
            assistant_msg = msg
            if assistant_msg.stop_reason in ("error", "aborted"):
                continue

            # Track tool calls from this assistant message
            tool_calls = [b for b in assistant_msg.content if b.type == "toolCall"]
            if tool_calls:
                pending_tool_calls = tool_calls  # type: ignore
                existing_tool_result_ids = set()

            result.append(msg)

        elif msg.role == "toolResult":
            existing_tool_result_ids.add(msg.tool_call_id)
            result.append(msg)

        elif msg.role == "user":
            # User message interrupts tool flow - insert synthetic results
            if pending_tool_calls:
                for tc in pending_tool_calls:
                    if tc.id not in existing_tool_result_ids:
                        result.append(
                            ToolResultMessage(
                                role="toolResult",
                                tool_call_id=tc.id,
                                tool_name=tc.name,
                                content=[TextContent(type="text", text="No result provided")],
                                is_error=True,
                                timestamp=int(time.time() * 1000),
                            )
                        )
                pending_tool_calls = []
                existing_tool_result_ids = set()
            result.append(msg)

        else:
            result.append(msg)

    return result
